### PR TITLE
Replace pybind11 with nanobind in frontend

### PR DESCRIPTION
**Pull request by joeycarter**
**Context:**

The Catalyst frontend contains a small Python extension module at [frontend/catalyst/utils/wrapper.cpp](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/catalyst/utils/wrapper.cpp), which is an importable Python module written in C/C++ used to wrap the entry point of compiled programs. The Python-C++ bindings were originally implemented using pybind11. This PR is part of a larger effort to replace all pybind11 code with nanobind.

**Description of the Change:**

This change replaces all the pybind11 code in the frontend with the equivalent nanobind objects and operations.

It was also necessary to modify the frontend build system to build the `wrapper` module using CMake for compatibility with nanobind, rather than in `setup.py` with the `intree_extensions` setuptools utility included with pybind11.

**Benefits:**

See Epic [68472](https://app.shortcut.com/xanaduai/epic/68472) for a list of nanobind's benefits.

-----

[[sc-72837](https://app.shortcut.com/xanaduai/story/72837/replace-pybind11-with-nanobind-in-the-frontend)]

### CODE REVIEW

**Comment by joeycarter** [comment]
The current implementation appears to work with one caveat: if `wrapper.wrap()` raises an exception, the type of the exception is lost and replaced with a generic `RuntimeError`. Consider the following example:

```python
import pennylane as qml
import jax.numpy as jnp
from catalyst.api_extensions.callbacks import base_callback


@base_callback
def identity(arg) -> int:
    return arg

@qml.qjit
def cir(x):
    return identity(x)

arg = jnp.array([0.0, 1.0, 2.0])
cir(arg)
```

The original behaviour was to raise a `TypeError` with the following stack trace:

```
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    cir(arg)
  File ".../lib/python3.12/site-packages/pennylane/logging/decorators.py", line 61, in wrapper_entry
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/catalyst/jit.py", line 469, in __call__
    return self.run(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/catalyst/debug/instruments.py", line 143, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/pennylane/logging/decorators.py", line 61, in wrapper_entry
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/catalyst/jit.py", line 693, in run
    results = self.compiled_function(*args, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/catalyst/compiled_functions.py", line 345, in __call__
    result = CompiledFunction._exec(
             ^^^^^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/catalyst/compiled_functions.py", line 166, in _exec
    retval = wrapper.wrap(lib.function, args, result_desc, lib.mem_transfer, numpy_dict)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/catalyst/api_extensions/callbacks.py", line 587, in __call__
    self._check_types(retval, exp_aval)
  File ".../lib/python3.12/site-packages/catalyst/api_extensions/callbacks.py", line 605, in _check_types
    raise TypeError(msg)
TypeError: Callback identity expected type ShapedArray(int64[]) but observed ShapedArray(float64[3]) in its return value
```

Using nanobind, the stack trace becomes:

```python
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    cir(arg)
  File ".../lib/python3.12/site-packages/pennylane/logging/decorators.py", line 61, in wrapper_entry
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File ".../catalyst/frontend/catalyst/jit.py", line 525, in __call__
    return self.run(args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^
  File ".../catalyst/frontend/catalyst/debug/instruments.py", line 143, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File ".../lib/python3.12/site-packages/pennylane/logging/decorators.py", line 61, in wrapper_entry
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File ".../catalyst/frontend/catalyst/jit.py", line 762, in run
    results = self.compiled_function(*args, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ".../catalyst/frontend/catalyst/compiled_functions.py", line 345, in __call__
    result = CompiledFunction._exec(
             ^^^^^^^^^^^^^^^^^^^^^^^
  File ".../catalyst/frontend/catalyst/compiled_functions.py", line 166, in _exec
    retval = wrapper.wrap(lib.function, args, result_desc, lib.mem_transfer, numpy_dict)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: TypeError: Callback identity expected type ShapedArray(int64[]) but observed ShapedArray(float64[3]) in its return value

At:
  .../catalyst/frontend/catalyst/api_extensions/callbacks.py(605): _check_types
  .../catalyst/frontend/catalyst/api_extensions/callbacks.py(587): __call__
  .../catalyst/frontend/catalyst/compiled_functions.py(166): _exec
  .../catalyst/frontend/catalyst/compiled_functions.py(345): __call__
  .../catalyst/frontend/catalyst/jit.py(762): run
  .../lib/python3.12/site-packages/pennylane/logging/decorators.py(61): wrapper_entry
  .../catalyst/frontend/catalyst/debug/instruments.py(143): wrapper
  .../catalyst/frontend/catalyst/jit.py(525): __call__
  .../lib/python3.12/site-packages/pennylane/logging/decorators.py(61): wrapper_entry
  .../catalyst/test.py(15): <module>
```

The stack traces and exceptions are equivalent, but still different. Moreover, the original trace is arguably in a more "natural" form and easier to read for someone familiar with Python. Also, since the exception type has changed, a few tests are failing, e.g. `test_incorrect_return()` from `frontend/test/pytest/test_callback.py`, where this example was drawn from.

I'm investigating now if it's possible to preserve the original stack trace and exception types when using nanobind.

**Comment by dime10** [comment]
Thank you @joeycarter, I'm okay with that! I think exceptions from Python through the C++ and back to Python might also not be too common anyways.

**Comment by erick-xanadu** [comment]
@joeycarter , can the `catch` statement be placed not in the wrapper, but in the runtime. E.g. wrapping `callbackCall` in `void __catalyst_inactive_callback(int64_t identifier, int64_t argc, int64_t retc, ...)`. This is because the call stack looks as follows (up means lower in the call stack)

1. python calling the wrapper pybind11 / nanobind wrap function
2. the ctypes function pointer
3. landing in `__catalyst_inactive_callback`
4. `callbackCall`
5. pybind11 / nanobind module for callbacks
6. back to python

Not sure if placing the `catch` statement closer to its source will yield the results that you want.

**Comment by joeycarter** [comment]
Thanks @dime10 and @erick-xanadu! I'll go ahead and change the tests to accept `RuntimeError`s, that's certainly the most straightforward approach.

I'm not sure at the moment if moving the catch block will help; ultimately I don't think we'll want a try-catch block here to process the errors in the wrapper, this was just to test what type of exception is being raised. I'll work on replacing pybind11 -> nanobind in the runtime in a future PR, and I can test that there too. If we can fix the issue there then we can always revert the tests back to accept the original exception types.

**Comment by joeycarter** [comment]
The CI build revealed a few issues with the new CMake configuration I've added for the frontend. Working on a fix now.

**Comment by joeycarter** [comment]
I was looking into the possibility of using nanobind's Stable ABI feature, described here: https://nanobind.readthedocs.io/en/latest/packaging.html#step-3-set-up-a-cmake-build-system. However, it seems this is incompatible with the NumPy C API. If I create a dummy module containing only the lines

```cpp
#define NPY_NO_DEPRECATED_API NPY_1_24_API_VERSION
#include "numpy/ndarrayobject.h"
```

I get a compiler error:

```
In file included from .../catalyst/frontend/catalyst/utils/wrapper.cpp:23:
In file included from .../lib/python3.12/site-packages/numpy/_core/include/numpy/ndarrayobject.h:13:
.../lib/python3.12/site-packages/numpy/_core/include/numpy/dtype_api.h:452:5: error: no matching function for call to 'Py_INCREF'
  452 |     Py_INCREF(o);
      |     ^~~~~~~~~
/usr/include/python3.12/object.h:624:37: note: candidate function not viable: cannot convert argument of incomplete type 'PyArray_DTypeMeta *' (aka '_typeobject *') to 'PyObject *' (aka '_object *') for 1st argument
  624 | static inline Py_ALWAYS_INLINE void Py_INCREF(PyObject *op)
      |                                     ^         ~~~~~~~~~~~~
```

Compiling with `STABLE_ABI` in the nanobind CMake function `nanobind_add_module` adds the macro definition `-DPy_LIMITED_API=0x030C0000` (which is the hex code for Python 3.12.0, the version I'm using). The issue is in the numpy header [numpy/_core/include/numpy/dtype_api.h](https://github.com/numpy/numpy/blob/5c8f540f3ed82138d7075fceb5e9738e918abf9d/numpy/_core/include/numpy/dtype_api.h#L18-L55); when `Py_LIMITED_API` is defined, the type `PyArray_DTypeMeta` is incomplete, hence the error.

This appears to be a bug with numpy and not something we can fix, which might block us from using nanobind's Stable ABI feature.

**Comment by dime10** [comment]
Does this not imply that it's possible to use the stable abi with numpy's c-api ? 
https://github.com/numpy/numpy/issues/26756

**Comment by joeycarter** [comment]
> Does this not imply that it's possible to use the stable abi with numpy's c-api ? [numpy/numpy#26756](https://github.com/numpy/numpy/issues/26756)

Thanks @dime10! Yes if we move to numpy>=2.0.1 then we'll be able to use the Stable ABI (just tested locally and it appears to work). The only caveat is that PennyLane currently has the requirement `numpy<=2.0` in its [setup.py](https://github.com/PennyLaneAI/pennylane/blob/master/setup.py#L24), so we'll need to request support for at least 2.0.1.

**Comment by dime10** [comment]
> > Does this not imply that it's possible to use the stable abi with numpy's c-api ? [numpy/numpy#26756](https://github.com/numpy/numpy/issues/26756)
> 
> Thanks @dime10! Yes if we move to numpy>=2.0.1 then we'll be able to use the Stable ABI (just tested locally and it appears to work). The only caveat is that PennyLane currently has the requirement `numpy<=2.0` in its [setup.py](https://github.com/PennyLaneAI/pennylane/blob/master/setup.py#L24), so we'll need to request support for at least 2.0.1.

Have you tried with a lower numpy version as well? The issue I think suggests that that is also possible. I'm asking because we don't want to force people to use numpy 2+ if we don't have to. Would `numpy!=2.0.0`be sufficient in that case?

**Comment by joeycarter** [comment]
> Have you tried with a lower numpy version as well? The issue I think suggests that that is also possible. I'm asking because we don't want to force people to use numpy 2+ if we don't have to. Would `numpy!=2.0.0`be sufficient in that case?

We discussed offline with @mlxd, we'll use `numpy<2.1,!=2.0.0`. I tested with numpy 1.26.4 as well and we can build and run Catalyst without issue.

I'll wait until https://github.com/PennyLaneAI/pennylane/pull/6342 is merged and then I think this PR will be ready for review.

File: pyproject.toml
Comment (mlxd) [inline pyproject.toml:42]:
```suggestion
requires = ["setuptools>=62", "wheel", "pybind11>=2.12.0", "numpy!=2.0.0", "nanobind", "cmake", "ninja"]
```

**Comment by joeycarter** [comment]
Praise be, the wheels are building :tada:

There's one pytest that's failing; I chatted with @paul0403 and confirmed that this is a non-deterministic failure with the result falling just outside the tolerance of the expected result. We should be able to fix this by seeding the qjit-ed circuit.

This PR is now ready for review. The new build system that uses CMake for the wrapper module is a little complex, since there is another Extension module, libcustom_calls, that still uses the old setuptools build system. We felt that updating the libcustom_calls module to use CMake as well was beyond the scope of this PR, but hopefully we can address it soon in a future PR.

**Comment by paul0403** [comment]
> We should be able to fix this by seeding the qjit-ed circuit.

Remember to open a new PR and seed that test :smirk:

**Comment by joeycarter** [comment]
> > We should be able to fix this by seeding the qjit-ed circuit.
> 
> Remember to open a new PR and seed that test 😏

PR is here: #1200 ;)

**Review by erick-xanadu — APPROVED** [review]
:100:

File: doc/requirements.txt
Comment (dime10) [inline doc/requirements.txt:25]: Is this a doc specific requirement? I'm just trying to think why it might show up here since we don't include any of our other build requirements (which is good, we wan't to avoid duplication whenever possible)

File: frontend/catalyst/utils/CMakeLists.txt
Comment (dime10) [inline frontend/catalyst/utils/CMakeLists.txt:1]: Nice, thanks for leaving comments what each command does!

File: frontend/catalyst/utils/wrapper.cpp
Comment (dime10) [inline frontend/catalyst/utils/wrapper.cpp:22]: Just curious about this change

File: frontend/catalyst/utils/wrapper.cpp
Comment (dime10) [inline frontend/catalyst/utils/wrapper.cpp:173]: Just curious if you could explain the use of borrow :)

File: frontend/catalyst/utils/wrapper.cpp
Comment (dime10) [inline frontend/catalyst/utils/wrapper.cpp:1]: Nice, good to know so far it's largely identical to pybind 💯

File: pyproject.toml
Comment (dime10) [inline pyproject.toml:42]: Do we still require pybind then? (or do you plan to remove all pybind installations after all PRs are merged?)

File: requirements.txt
Comment (dime10) [inline requirements.txt:8]: Could you comment on this requirement? For example, in the pyproject.toml there is no version lower bound.

**Review by dime10 — COMMENTED** [review]
Nice work 💯

File: requirements.txt
Comment (dime10) [reply requirements.txt:8]: Just another comment on the strategy we have regarding requirements / dependencies.

- `pyproject.toml`: This is the new to specify _build_ requirements for python packages, so anything that's needed in order to build our python package (i.e. the frontend) can be put in there, and it will automatically be installed in a separate environment only for the duration of the build process.
- `requirements.txt`: Since we have other components that are not Python packages (mlir, runtime), we put their _build_ requirements in this file (e.g. numpy is needed to build mlir) and tell users to install these packages before starting the build process, so this method is a little less automated than the first one. We also put other useful development tools in here like linters, formatters, testing requirements, etc. that we would developers to have.
- `setup.py`: Here we put any _runtime_ dependencies that the built Catalyst package has, enforcing their installation when a user installs Catalyst.
- documentation: A few additional requirements that are difficult to handle automatically are just referenced in the documentation, like installing a compiler.
- `doc/requirements.txt` & `benchmark/requirements.txt`: optional dependencies only used for those particular features

Ideally, we want to avoid putting the same requirement in more than one place (unless needed), and prefer "automatic" over "manual" methods.
In this case looking at nanobind for example, we'll need it for the frontend and the runtime, so putting it on both the pyproject.toml and requirements.txt is probably needed (although technically this PR might only need it in the pyproject), but I just wanted to give a rundown of the dependency handling :)

File: frontend/catalyst/utils/wrapper.cpp
Comment (joeycarter) [reply frontend/catalyst/utils/wrapper.cpp:22]: This was flagged by my [clangd `#include` manager](https://clangd.llvm.org/guides/include-cleaner), which

> follows the [include-what-you-use](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md) model: each source file should #include headers that declare the symbols it references, and no others.

[`numpy/arrayobject.h`](https://github.com/numpy/numpy/blob/v2.0.1/numpy/_core/include/numpy/arrayobject.h) appears to be a sort of alias for `"ndarrayobject.h"` in numpy 2.x. In numpy 1.x, [`numpy/arrayobject.h`](https://github.com/numpy/numpy/blob/v1.26.5/numpy/core/include/numpy/arrayobject.h) also includes `"npy_interrupt.h"`; we don't use the symbols defined in this header which is why the clangd tool complained.

I also found in the NumPy documentation that including the `numpy/ndarrayobject.h` header is preferred: https://numpy.org/devdocs/reference/c-api/array.html#including-and-importing-the-c-api

File: frontend/catalyst/utils/wrapper.cpp
Comment (joeycarter) [reply frontend/catalyst/utils/wrapper.cpp:1]: For the most part, yes it was a pretty easy transition (the changes to the build system notwithstanding)! :rocket:

File: frontend/catalyst/utils/wrapper.cpp
Comment (joeycarter) [reply frontend/catalyst/utils/wrapper.cpp:173]: I'm glad you pointed this out because I think that we should actually use [`nb::steal`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I0EN8nanobind5stealE1T6handle) here, not [`nb::borrow`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I0EN8nanobind6borrowE1T6handle).

The short answer is that without one of `nb::borrow` or `nb::steal`, you get a `std::bad_cast` runtime error!

Here, `new_array` is a raw [`PyObject*`](https://docs.python.org/3/c-api/structures.html#c.PyObject) pointer and `returns` is a [`nb::list`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4N8nanobind4listE) (a wrapper class representing a Python `list`). In order to call [`append()`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I0EN8nanobind4list6appendEvRR1T) on `returns`, its argument must be a wrapped Python object. To wrap the `PyObject*` pointer, we have the choice of `nb::steal`, which *steals* a reference to the object (and does **not** increment its reference count), or `nb::borrow` which *borrows* a reference to the object (and **does** increments its reference count). Stealing is generally for C API calls that return a new object that we now own and must manage, and borrowing is for C API calls that return a (borrowed) reference to an object that another part of the program owns and manages.

The documentation for [`PyArray_NewFromDescr`](https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_NewFromDescr) says that it steals a reference to `descr` (which makes sense since it's returning a *new* array), so we should use `nb::steal` here.

File: pyproject.toml
Comment (joeycarter) [reply pyproject.toml:42]: Yes I was planning to remove all references to pybind11 in a separate PR once all of the other PRs are merged.

File: frontend/catalyst/utils/wrapper.cpp
Comment (dime10) [reply frontend/catalyst/utils/wrapper.cpp:22]: Oh neat, I've looking for C++ include managers

File: frontend/catalyst/utils/wrapper.cpp
Comment (dime10) [reply frontend/catalyst/utils/wrapper.cpp:173]: Nice, thanks for explaining it :)

@erick-xanadu Does that makes sense to you since you wrote the original?

Comment (erick-xanadu) [reply frontend/catalyst/utils/wrapper.cpp:173]: I think borrow makes sense given that we manually decrement the reference in line 191. Have you tried stealing and everything worked correctly?

Comment (erick-xanadu) [reply frontend/catalyst/utils/wrapper.cpp:173]: If it works with borrow, then that should be enough proof too. You can manually [look at the reference count](https://docs.python.org/3/c-api/refcounting.html#c.Py_REFCNT). Essentially at the end of this method, the reference count should be two. One for the list and one for the dictionary.

Comment (dime10) [reply frontend/catalyst/utils/wrapper.cpp:173]: Ah, so we could either borrow and decrement, or steal?

Comment (erick-xanadu) [reply frontend/catalyst/utils/wrapper.cpp:173]: > Ah, so we could either borrow and decrement, or steal?

I think we need to `(borrow and decrement) xor steal`

Comment (joeycarter) [reply frontend/catalyst/utils/wrapper.cpp:173]: Thanks @dime10 and @erick-xanadu! I hadn't considered that we insert `new_array` in two places. I printed out the reference counts of `new_array` just to be sure; I get:

1. Increment ref count to 1 after array creation
2. Increment ref count to 2 after `returns.append(nb::borrow(new_array))`
3. Increment ref count to 3 after `numpy_arrays[pyLong] = nb::borrow(new_array)`
4. Decrement ref count to 2 with `Py_DECREF(new_array)`

If we were using the array once it might make more sense to use `nb::steal` so we don't have to manually decrement the ref count. We could steal once and borrow once, but that might looks strange and confuse anyone looking at this in the future. I propose we keep it as is with two borrows and a final `Py_DECREF`. I'll add a comment saying that the ref count of `new_array` at the end should be 2.

File: doc/requirements.txt
Comment (joeycarter) [reply doc/requirements.txt]: Good catch! I think I had added this in when trying to get the docs build to work, but the real issue was something else. I removed nanobind from this list (in 1333f83) and the docs still build, so indeed it wasn't necessary. Thanks!

**Comment by joeycarter** [comment]
I've added the "Do not merge" label; we'll merge this in along with the other nanobind PRs after the v0.9.0 release.

**Pull request merged**
